guacamole-common-js/src/main/webapp/modules/KeyEventInterpreter.js

```diff
@@ -0,0 +1,579 @@
+     * key is currently pressed. If a keysym has no entry in this map at all,
+     * it means that the key is not being pressed. Note that not all keysyms
+     * are necessarily tracked within this map - only those that are explicitly
+     * tracked.
+     *
+     * @private
+     * @type {Object.<String,KeyDefinition> }
+     */
+    var pressedKeys = {};
+
+    /**
+     * The current key event batch, containing a representation of all key
+     * events processed since the end of the last batch passed to onbatch.
+     * Null if no key events have been processed yet.
+     *
+     * @private
+     * @type {!KeyEventBatch}
+     */
+    var currentBatch = null;
+
+    /**
+     * The timestamp of the most recent key event processed.
+     *
+     * @private
+     * @type {Number}
+     */
+    var lastKeyEvent = 0;
+
+    /**
+     * Returns true if the currently-pressed keys are part of a shortcut, or
+     * false otherwise.
+     *
+     * @private
+     * @returns {!boolean}
+     *     True if the currently-pressed keys are part of a shortcut, or false
+     *     otherwise.
+     */
+    function isShortcut() {
+
+        // If one of the currently-pressed keys is non-printable, a shortcut
+        // is being typed
+        for (var keysym in pressedKeys) {
+            if (pressedKeys[keysym].value === null)
+                return true;
+        }
+
+        return false;
+    }
+
+    /**
+     * If the provided keysym corresponds to a valid UTF-8 character, return
+     * a KeyDefinition for that keysym. Otherwise, return null.
+     *
+     * @private
+     * @param {Number} keysym
+     *     The keysym to produce a UTF-8 KeyDefinition for, if valid.
+     *
+     * @returns {KeyDefinition}
+     *     A KeyDefinition for the provided keysym, if it's a valid UTF-8
+     *     keysym, or null otherwise.
+     */
+    function getUnicodeKeyDefinition(keysym) {
+
+        // Translate only if keysym maps to Unicode
+        if (keysym < 0x00 || (keysym > 0xFF && (keysym | 0xFFFF) != 0x0100FFFF))
+            return null;
+
+        var codepoint = keysym & 0xFFFF;
+        var mask;
+        var bytes;
+
+        // Determine size and initial byte mask
+        if (codepoint <= 0x007F) {
+            mask  = 0x00;
+            bytes = 1;
+        }
+        else if (codepoint <= 0x7FF) {
+            mask  = 0xC0;
+            bytes = 2;
+        }
+        else {
+            mask  = 0xE0;
+            bytes = 3;
+        }
+
+        var byteArray = new ArrayBuffer(bytes);
+        var byteView = new Int8Array(byteArray);
+
+        // Add trailing bytes, if any
+        for (var i = 1; i < bytes; i++) {
+            byteView[bytes - i] = 0x80 | (codepoint & 0x3F);
+            codepoint >>= 6;
+        }
+
+        // Set initial byte
+        byteView[0] = mask | codepoint;
+
+        // Convert to UTF8 string
+        var name = new TextDecoder("utf-8").decode(byteArray);
```

Comment:
Can you clarify why this is necessary? I read:

1. X11 keysym is converted to a Unicode codepoint.
2. Codepoint is manually encoded as UTF-8 bytes.
3. UTF-8 bytes are decoded to text.

Seems like a verrrry roundabout way to do:

```js
var name = String.fromCodePoint(codepoint);
```

Reply:
Or even:

```js
var name = String.fromCharCode(codepoint);
```

so long as we know `codepoint` will always be no greater than `0xFFFF`.

Reply:
Uh yeah, we definitely know that it will be no greater than `0xFFFF` considering `var codepoint = keysym & 0xFFFF;` on line 281. I'll use `fromCharCode`. All this stuff was carried over from the `guaclog` codebase and is indeed not needed.

Good catch.